cmake/package_version_template.h.in

```diff
@@ -0,0 +1,88 @@
+#ifndef @CMAKE_PROJECT_NAME@_VERSION_H
+#define @CMAKE_PROJECT_NAME@_VERSION_H
+
+#include <cstdint>
+#include <tuple>
+#include <ostream>
+#if __cplusplus >= 202002L
+#include <compare>
+#endif
+
+// Some preprocessor constants and macros for the use cases where they might be
+// necessary
+
+/// Define a version to be used in @CMAKE_PROJECT_NAME@.
+#define @CMAKE_PROJECT_NAME@_VERSION(major, minor, patch) (((unsigned long)(major) << 32) | ((unsigned long)(minor) << 16) | ((unsigned long)(patch)))
+/// Get the major version from a preprocessor defined version
+#define @CMAKE_PROJECT_NAME@_MAJOR_VERSION(v) (((v) & (-1UL >> 16)) >> 32)
+/// Get the minor version from a preprocessor defined version
+#define @CMAKE_PROJECT_NAME@_MINOR_VERSION(v) (((v) & (-1UL >> 32)) >> 16)
+/// Get the patch version from a preprocessor defined version
+#define @CMAKE_PROJECT_NAME@_PATCH_VERSION(v) ((v) & (-1UL >> 48))
+
+// Some helper constants that are populated by the cmake configure step
+#define @CMAKE_PROJECT_NAME@_VERSION_MAJOR @PACKAGE_VERSION_MAJOR@
+#define @CMAKE_PROJECT_NAME@_VERSION_MINOR @PACKAGE_VERSION_MINOR@
+#define @CMAKE_PROJECT_NAME@_VERSION_PATCH @PACKAGE_VERSION_PATCH@
+#define @CMAKE_PROJECT_NAME@_VERSION @PACKAGE_VERSION@
+
+/// The encoded version with which @CMAKE_PROJECT_NAME@ has been built
+#define @CMAKE_PROJECT_NAME@_BUILD_VERSION @CMAKE_PROJECT_NAME@_VERSION(@PACKAGE_VERSION_MAJOR@, @PACKAGE_VERSION_MINOR@, @PACKAGE_VERSION_PATCH@)
+
+namespace @CMAKE_PROJECT_NAME@::version {
+
+  /**
+   * Version class consisting of 3 16 bit unsigned integers to hold the major,
+   * minor and patch version. Provides constexpr comparison operators that allow
+   * to use this class in constexpr-if clauses.
+   */
+  struct Version {
```

Comment:
Similarly this class could live in a "global" header and general namespace and simply be included. In the end only the `constexpr Version build_version{@PACKAGE_VERSION_MAJOR@, @PACKAGE_VERSION_MINOR@, @PACKAGE_VERSION_PATCH@}` is really package specific.